File: src/audioWfc/constraints/PerfectCadenceSoftConstraint.java
package audioWfc.constraints;

import audioWfc.constraints.grabbers.ChordQualityConstantGrabber;
import audioWfc.constraints.grabbers.Grabber;
import audioWfc.constraints.grabbers.IntegerConstantGrabber;
import audioWfc.musicTheory.Key;
import audioWfc.musicTheory.chords.ChordQuality;

public class PerfectCadenceSoftConstraint extends GenericCadenceSoftConstraint{

    public PerfectCadenceSoftConstraint(double factor, Grabber<Key> grabber){
        super(factor, grabber, new IntegerConstantGrabber(7), new ChordQualityConstantGrabber(ChordQuality.MAJOR),
                new IntegerConstantGrabber(0), new ChordQualityConstantGrabber(ChordQuality.MAJOR));
    }


}

File: src/wfc/constraints/cadences/PerfectCadenceSoftConstraint.ts
import { ConcreteChordQuality } from "../../../music_theory/Chord"
import { GenericCadenceSoftConstraint } from "./GenericCadenceSoftConstraint"
import { Grabber } from "../../Grabber"
import { MusicalKey } from "../../../music_theory/MusicalKey"
import { chordConstraintTypeToName } from "../constraintUtils"
import { constantConcreteChordQualityGrabber, constantNumberGrabber } from "../../grabbers/constantGrabbers"

export const PerfectCadenceSoftConstraintInit = {
	type: "PerfectCadenceSoftConstraint" as const,
	bonus: 1,
	validByDefault: true as const,
	reachOverPrev: true,
	reachOverNext: true
}

export type PerfectCadenceSoftConstraintIR =
	typeof PerfectCadenceSoftConstraintInit
export class PerfectCadenceSoftConstraint extends GenericCadenceSoftConstraint {
	name = chordConstraintTypeToName.get(
		PerfectCadenceSoftConstraintInit.type,
	)!.name as string
	constructor(bonus: number, keyGrabber: Grabber<MusicalKey>, reachOverPrev: boolean, reachOverNext: boolean) {
		super(
			bonus,
			constantNumberGrabber(7),
			constantNumberGrabber(0),
			constantConcreteChordQualityGrabber("major" as ConcreteChordQuality),
			constantConcreteChordQualityGrabber("major" as ConcreteChordQuality),
			keyGrabber,
			reachOverPrev,
			reachOverNext
		)
	}
}
